### PR TITLE
UGENE-7078 Set test directory from UGENE_TESTS_PATH if possible for C…

### PR DESCRIPTION
**Pull request by DmitriiSukhomlinov**
…InterfaceSasTests

https://ugene.dev/tracker/browse/UGENE-7078

### CODE REVIEW

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 1]
```diff
@@ -31,12 +32,32 @@
...
-#else
-static const QString PROPER_WD_SCHEMES_PATH = WORKING_DIR + "/../../../../../../test/_common_data/cmdline/wd-sas-schemes/";
-#endif
+#include <QFileInfo>
+
+static QString getTestDirImpl() {
+    QString testDir = qgetenv("UGENE_TESTS_PATH");
+    auto getDefaultTestDir = []() -> QString {
+        QString defaultTestDir;
+        #ifndef Q_OS_MAC
```
Comment (yalgaer) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: nit: we have more readable isOsMac()/isOsLinux() methods.

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 2]
```diff
@@ -31,12 +32,32 @@
...
+        defaultTestDir = U2::AppContext::getWorkingDirectoryPath() + "/../../../../../../test/";
+        #endif
+        return defaultTestDir;
+    };
+    if (testDir.isEmpty()) {
+        testDir = getDefaultTestDir();
+    } else if (!QFileInfo::exists(testDir)) {
+        QString defaultTestDir = getDefaultTestDir();
+        U2::coreLog.error(QString("UGENE_TESTS_PATH doesn't exist: '%1'. The default path is set: '%2'.").arg(testDir).arg(defaultTestDir));
+        testDir = getDefaultTestDir();
```
Comment (yalgaer) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: testDir = defaultTestDir

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 3]
```diff
@@ -31,12 +32,32 @@
...
+    }
+
+    testDir = testDir + (testDir.endsWith("/") ? "" : "/");
+    return testDir;
+}
+
+static const QString PROPER_WD_SCHEMES_PATH = getTestDirImpl() + "_common_data/cmdline/wd-sas-schemes/";
```
Comment (yalgaer) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: Looks like this variable is not used at all.
If it is still needed why is it called 'PROPER' do we plan to have 'IMPROPER' ones?

**Review by yalgaer — CHANGES_REQUESTED** [review]

**Review by atiunov — DISMISSED** [review]

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 4]
```diff
@@ -31,12 +32,32 @@
...
+        #ifndef Q_OS_MAC
+        defaultTestDir = U2::AppContext::getWorkingDirectoryPath() + "/../../test/";
+        #else
+        defaultTestDir = U2::AppContext::getWorkingDirectoryPath() + "/../../../../../../test/";
+        #endif
+        return defaultTestDir;
+    };
+    if (testDir.isEmpty()) {
+        testDir = getDefaultTestDir();
+    } else if (!QFileInfo::exists(testDir)) {
```
Comment (evelinabis) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp:48]: Is it possible to create a directory if it doesn't exist? Or doesn't it make sense?

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 5]
```diff
@@ -31,12 +32,32 @@
...
+        QString defaultTestDir = getDefaultTestDir();
+        U2::coreLog.error(QString("UGENE_TESTS_PATH doesn't exist: '%1'. The default path is set: '%2'.").arg(testDir).arg(defaultTestDir));
+        testDir = getDefaultTestDir();
+    }
+
+    testDir = testDir + (testDir.endsWith("/") ? "" : "/");
+    return testDir;
```
Comment (evelinabis) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp:54]: Do you need to check that the directory has the appropriate permissions to work with it (read/write)?

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 1 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: Fixed

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 4 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp:48]: The "UGENE_TESTS_PATH" using supposes that this directory already exists

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 2 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: Fixed

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 5 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp:54]: This is for testing,so this directory is set up by us. I think, permission check is excess

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 3 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: I don't get. This variable is used in many places CInterfaceSasTests.cpp. 

I have renamed it.

**Comment by DmitriiSukhomlinov** [comment]
Check again

**Review by evelinabis — DISMISSED** [review]

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp [hunk 6]
```diff
@@ -31,12 +33,32 @@
...
-#ifndef Q_OS_MAC
-static const QString PROPER_WD_SCHEMES_PATH = WORKING_DIR + "/../../test/_common_data/cmdline/wd-sas-schemes/";
-#else
-static const QString PROPER_WD_SCHEMES_PATH = WORKING_DIR + "/../../../../../../test/_common_data/cmdline/wd-sas-schemes/";
-#endif
+#include <QFileInfo>
+
+static QString getTestDirImpl() {
+    QString testDir = qgetenv("UGENE_TESTS_PATH");
+    auto getDefaultTestDir = []() -> QString {
```
Comment (yalgaer) [inline src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: Why do you want to have a long lambda here and not use a trivial code like
QString defaultTestDir = isOsMac() ?  "dir1" : "dir2" ?

**Review by yalgaer — CHANGES_REQUESTED** [review]

File: src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp (hunk 6 above)
Comment (DmitriiSukhomlinov) [reply src/plugins/api_tests/src/core/external_script/base_scheme_interface/CInterfaceSasTests.cpp]: Fixed

**Review by evelinabis — APPROVED** [review]

**Comment by DmitriiSukhomlinov** [comment]
Tests have passed correctly on [Windows](http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/viewLog.html?buildId=213524&tab=buildResultsDiv&buildTypeId=bt117), [Linux](http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/viewLog.html?buildId=213520&tab=buildResultsDiv&buildTypeId=bt50) and [macOS](http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/viewLog.html?buildId=213587&tab=buildResultsDiv&buildTypeId=UgeneNightlyBuilds_NightlyMacOsX1075macX64)

**Review by yalgaer — APPROVED** [review]

**Pull request merged**